### DISCUSSION TITLE
about sparse embedding, may upper_bound is calc with max_value * query_value on this dim, instead only max_value on this dim?

### DISCUSSION THREAD

**Discussion by ldak4747** [milvus-io/milvus#36717] [Ideas & Feature requests]
in src/index/sparse/sparse_inverted_index.h InvertedIndex::search_wand, when calc the upper_bound by traverse cursors, codes as "upper_bound += cursors[pivot]->max_score();", may the codes could be "upper_bound += (cursors[pivot]->max_score() * query_value);" (pseudocode)？

for example, index_value on one dim is from 0 to 1, but query_value on one dim is 1e10, so the inner product maybe very huge on this dim, the distance between 0.1 * 1e10 and 0.9 * 1e10 maybe very huge, but distance between 0.1 and 0.9 maybe no influence to prevent found_pivot = false

my idea is:
1. upper_bound is a accumulated sum with every dim
2. but the inner product is query_value * index_value at every dim
3. so if upper_bound is composed by only  index_value at every dim, maybe not precise?

**Comment by ldak4747** [comment]
另外有一个问题想请赐教：
wand算法，看起来是鉴于pivot之前的dim，因其max_value之和并未大于堆顶，所以认为它们可能不特别"重要"，但为什么只是把，pivot之前的仅一个cursor，做doc_id的"skip"，而不是pivot之前的全部cursor，都做skip，也就是，都去忽略掉低于cursors[pivot]首doc_id的所有doc_ids？

也就是重点是，只有第一个首doc_id不等于cursors[pivot]首doc_id的cursor，才会做skip？
伪代码类似如下：
            size_t next_list = pivot;
            while (cursors[next_list].InternalId() == pivot_id) {
                --next_list;
            }

            cursors[next_list].Seek(pivot_id); // 为什么不是全部pivot之前，首doc_id不等于cursors[pivot]首doc_id的cursors，都做skip的操作？

**Comment by zhengbuqian** [comment]
Hi @ldak4747,

Regarding your first question: as mentioned in my reply in https://github.com/milvus-io/milvus/issues/36711, "query_value is multiplied onto max_score when the cursor is created.". So no need to multiply it again.

Regarding the second question: it is only a choice of implementation, not correctness. Possible strategies could be:

1. advance any of the cursors
2. advance all the cursors
3. advance the cursor that'll produce the most skips.

Any strategy is ok.

**Reply by ldak4747** [reply]
1. about the issue，OK！
2. about the second questions，I guess you had verified that the choice now(only do skip for the first cursor with different first-doc_id) maybe the best choice, is it right?
